### ISSUE TITLE
Disallow objects with extra properties from being assigned to wider types

### ISSUE THREAD

**Issue by twilkinson3421** [microsoft/TypeScript#62898]
### 🔍 Search Terms

"extra properties", "narrow erased", "track narrowed"

### ✅ Viability Checklist

- [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [x] This wouldn't change the runtime behavior of existing JavaScript code
- [x] This could be implemented without emitting different JS based on the types of the expressions
- [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [x] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [x] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

Have some way of remembering when a variable could have extra properties not present in its type. Such variables should not be allowed to be assigned to wider types, as this can lead to variables being incorrectly typed.

### 📃 Motivating Example

The following example results in `h` being incorrectly typed, yet no error is reported.

```typescript
type A = { 0: "a"; 1: "b"; 2: "c" };

const n = { 0: "a", 1: "b", 2: "c", length: 3 } as const;

const g: A = n;

type B = { 0: "a"; 1: "b"; 2: "c"; length?: boolean };

const h: B = g;
```

If we could somehow remember that `g` might have extra properties that aren't defined in its type, `A`, then it could be disallowed from being assigned to a wider type (one whose properties are a superset of `A`), such as `B`.

### 💻 Use Cases

1. What do you want to use this for?
Having variables be typed correctly
2. What shortcomings exist with current approaches?
Variables can be incorrectly typed
3. What workarounds are you using in the meantime?
None

**Comment by nmain** [comment]
If this was applied to all existing types, it would be a massive breaking change.  If it was applied only to specific opt-in types, it'd be https://github.com/microsoft/TypeScript/issues/12936.
